### PR TITLE
Process implement

### PR DESCRIPTION
**Pull request by tymefighter**

### CODE REVIEW

File: Benchmark/System/Process.hs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import System.IO (openFile, hClose, IOMode(..))
+
+import Control.Monad (replicateM_)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+
+import Gauge
+
+runCatCmd :: IO ()
+runCatCmd = do
```
[the remark this answers is not in the document]
Comment (tymefighter) [reply Benchmark/System/Process.hs:29]: ok

File: streamly-process.cabal [hunk 2]
```diff
@@ -0,0 +1,63 @@
+cabal-version:       >=1.10
+-- Initial package description 'streamly-process.cabal' generated by 'cabal
```
Comment (harendra-kumar) [inline streamly-process.cabal:2]: remove this.

File: src/Streamly/System/Process.hs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+withExe :: 
+        (IsStream t, MonadCatch m, MonadIO m)
+        => FilePath          -- ^ Path to Executable
+        -> [String]          -- ^ Arguments
+        -> (Handle -> t m a) 
+        -- ^ Given handle to read from output of 
+        -- process generate stream output
+        -> t m a             -- ^ Stream produced
+withExe fpath args genStrm = S.bracket pre post body
+    where
```
Comment (harendra-kumar) [inline src/Streamly/System/Process.hs:87]: always add a blank line before where.

File: src/Streamly/System/Process.hs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+openProcInp ::  
+    FilePath                                -- ^ Path to Executable
+    -> [String]                             -- ^ Arguments
+    -> IO (Handle, Handle, ProcessHandle)   
+    -- ^ (Input Handle, Output Handle, Process Handle)
+openProcInp fpath args = do
+    (readInpEnd, writeInpEnd) <- createPipe
+    (readOutEnd, writeOutEnd) <- createPipe
+    let 
+        procObj = (proc fpath args) {
```
Comment (harendra-kumar) [inline src/Streamly/System/Process.hs:106]: Combine let with the next line `let procObj = (proc fpath args) {`

File: test/Streamly/System/Process.hs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+byteFile :: String
+byteFile = "./byteFile"
+
+charFile :: String
+charFile = "./largeCharFile"
+
+generateByteFile :: Int -> IO ()
+generateByteFile blockCount = 
+    let
```
Comment (harendra-kumar) [inline test/Streamly/System/Process.hs:109]: you can move this let inside do block. Also, combine the first line of `let` block with the `let` keyword. Apply the same to all code, in benchmarks as well.

File: Benchmark/System/Process.hs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    where 
+    
+    envIO = do
+        outputHdl <- openFile devNull WriteMode
+        newIORef outputHdl
+
+    closeHandle ioRefOutputHdl = do
+        outputHdl <- readIORef ioRefOutputHdl
+        hClose outputHdl
+        writeIORef ioRefOutputHdl outputHdl
```
Comment (pranaysashank) [inline Benchmark/System/Process.hs:116]: What is the point in writing this handle to the IORef again?

File: Benchmark/System/Process.hs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    envIO = do 
+        inputHdl <- openFile largeCharFile ReadMode
+        outputHdl <- openFile devNull WriteMode
+        newIORef (inputHdl, outputHdl)
+
+    closeHandles ioRefHandles = do
+        (inputHdl, outputHdl) <- readIORef ioRefHandles
+        hClose inputHdl
+        hClose outputHdl
+        writeIORef ioRefHandles (inputHdl, outputHdl)
```
Comment (pranaysashank) [inline Benchmark/System/Process.hs:132]: Refer to the above comment

File: src/Streamly/System/Process.hs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    -> [String]             -- ^ Arguments
+    -> t m Word8            -- ^ Input stream to the process
+    -> (Handle -> t m a)    
+    -- ^ Handle to read output of process and generate stream
+    -> t m a                -- ^ Stream produced
+withInpExe fpath args input genStrm = S.bracket pre post body
+
+    where
+        
+        writeAction writeHdl =
```
Comment (harendra-kumar) [inline src/Streamly/System/Process.hs:120]: reduce indentation, align with `where`.

File: cabal.project [hunk 9]
```diff
@@ -0,0 +1,5 @@
+packages: .
+source-repository-package
+    type: git
+    location: https://github.com/composewell/streamly
+    tag: master
```
[the remark this answers is not in the document]
Comment (tymefighter) [reply cabal.project:5]: ok

File: streamly-process.cabal [hunk 10]
```diff
@@ -0,0 +1,70 @@
+cabal-version:       2.2
+name:                streamly-process
+version:             0.1.0.0
+synopsis:           Beautiful Streaming, Concurrent and Reactive Composition
```
[the remark this answers is not in the document]
Comment (tymefighter) [reply streamly-process.cabal:4]: what do I write ?

**Comment by tymefighter** [comment]
We have to write CI for this package

**Pull request merged**